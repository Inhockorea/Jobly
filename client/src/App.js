/** App: Component that renders home page with description of Jobly and depending
 *  on token status, either a Login button (if it doesn't exist), or a 'welcome back' message
 *    - Holds state of token and username
 *    - Provider of UserContext is located here
 *    - Used in Index component
 *    - Uses Routes and Navigation components
 */


import React, { useEffect, useState } from "react";
import { Redirect } from "react-router-dom";
import jwt from "jsonwebtoken";

import Routes from "./nav-routes/Routes";
import Navigation from "./nav-routes/Navigation";
import JoblyApi from "./api/JoblyApi";
import UserContext from "./auth/UserContext";
import LoadingSpinner from "./shared/LoadingSpinner";


function App() {
  const [token, setToken] = useState(localStorage.getItem('token'));
  const [currentUser, setCurrentUser] = useState(null);
  const [requestCompleted, setRequestCompleted] = useState(false);

  /** handleLogout: logs out current user by removing token and resetting user context to null */

  function handleLogout() {
    localStorage.removeItem('token');
    setCurrentUser(null);
    return <Redirect to='/' />
  }

  /** Sets the current user in state from the username in the token
   *  If token is null, set the current user to null to logout
   */

  useEffect(() => {
    if (token !== null) {
      localStorage.setItem('token', token);
      const { username } = jwt.decode(token);

      async function fetchUser() {
        try {
          let resp = await JoblyApi.request(`users/${username}`);
          setCurrentUser(resp.user);
          // waits until user has been set to change requestCompleted state to true
          setRequestCompleted(true);
        } catch (err) {
          console.error(err);
        }
      }
      fetchUser();
    } else {
      setCurrentUser(null);
      setRequestCompleted(true);
    }
  }, [token]);

  function returnRoutes() {
    if (requestCompleted !== true) {
      return <LoadingSpinner />
    } else {
      return <Routes login={login} signup={signup} />
    }
  }

  /** login: site-wide login */

  async function login(loginData) {
    let { username, password } = loginData;
    try {
      let resp = await JoblyApi.request('login', { username, password }, "post");
      setToken(resp.token);
      return { success: true }
    } catch (err) {
      return { success: false, errors: err }
    }
  }

  /** signup: site-wide user signup */

  async function signup(signupData) {
    try {
      let resp = await JoblyApi.request('users', signupData, "post");
      setToken(resp.token);
      return { success: true }
    } catch (err) {
      return { success: false, errors: err }
    }
  }


  return (
    <div>
      <UserContext.Provider value={{ currentUser, setCurrentUser }}>
        <Navigation handleLogout={handleLogout} />
        {returnRoutes()}
      </UserContext.Provider>
    </div >
  );
}

export default App;