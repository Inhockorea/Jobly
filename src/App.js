import React, { useEffect, useState } from "react";
import { BrowserRouter } from "react-router-dom";
import jwt from "jsonwebtoken";

import Routes from "./Routes";
import Navigation from "./Navigation";
import TokenContext from "./tokenContext";


/** App: Component that renders home page with description of Jobly and depending 
 *  on token status, either a Login button (if it doesn't exist), or a 'welcome back' message
 *    - Holds state of token and username
 *    - Provider of TokenContext is located here
 *    - Used in Index component
 *    - Uses Routes and Navigation components
 */

function App() {
  const [token, setToken] = useState(localStorage.getItem('token'));
  const [username, setUsername] = useState(null);

  useEffect(() => {
    localStorage.setItem('token', token);

    if (token !== "null") {
      let payload = jwt.decode(token);
      let updatedUsername = payload.username;
      setUsername(updatedUsername);
    }
  }, [token]);


return (
  <div>
    <TokenContext.Provider value={{ token, setToken, username }}>
      <BrowserRouter>
        <Navigation />
        <div className="container">
          <Routes />
        </div>
      </BrowserRouter>
    </TokenContext.Provider>
    </div >
  );
}

export default App;