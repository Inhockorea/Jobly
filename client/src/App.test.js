import React from 'react';
import { render, waitForElement } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import App from './App';


describe("Jobly routes", function() {
  it("renders without crashing", async function() {
    const { getByText } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    await waitForElement(() => getByText("Log in"));
  });

  it("renders the home snapshot", async function() {
    const { asFragment, getByText, getAllByText } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    await waitForElement(() => getByText("All the jobs in one convenient place."));
    getByText("Login");
    getAllByText("Jobly");

    expect(asFragment()).toMatchSnapshot();
  });
});