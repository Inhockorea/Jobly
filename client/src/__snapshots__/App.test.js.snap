// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jobly routes renders the home snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <nav
        class="Navigation navbar navbar-expand-lg navbar-light bg-light"
      >
        <a
          aria-current="page"
          class="navbar-brand active"
          href="/"
        >
          Jobly
        </a>
        <ul
          class="navbar-nav ml-auto"
        >
          <li
            class="Navigation nav-item"
          >
            <a
              href="/login"
            >
              Login
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <div
      class="Home"
    >
      <h1>
        Jobly
      </h1>
      <p>
        All the jobs in one convenient place.
      </p>
      <a
        href="/login"
      >
        <button
          class="btn btn-primary home-login-button"
        >
          Log in
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
